src/conn/conn_compact.c

```diff
@@ -21,21 +22,256 @@ __compact_server_run_chk(WT_SESSION_IMPL *session)
+         * dropped and recreated this table. Reset the entry in this case.
+         */
+        if (id != compact_stat->id) {
+            __background_compact_list_remove(session, compact_stat);
+            return (NULL);
+        }
+
+        return (compact_stat);
+    }
+
+    return (NULL);
+}
+
+/*
+ * __background_compact_should_run --
+ *     Check whether we should proceed with calling compaction on the given file.
+ */
+static bool
+__background_compact_should_run(WT_SESSION_IMPL *session, const char *uri, int64_t id)
+{
+    WT_BACKGROUND_COMPACT_STAT *compact_stat;
+    WT_CONNECTION_IMPL *conn;
+    uint64_t cur_time;
+
+    conn = S2C(session);
+
+    /* The history store file should not be compacted. */
+    if (WT_STREQ(uri, WT_HS_URI))
+        return (false);
+
+    /* If we haven't seen this file before we should try and compact it. */
+    compact_stat = __background_compact_get_stat(session, uri, id);
+    if (compact_stat == NULL)
+        return (true);
+
+    /* If we have been unsuccessful recently skip this file for some time. */
+    cur_time = __wt_clock(session);
+    if (!compact_stat->prev_compact_success &&
+      WT_CLOCKDIFF_SEC(cur_time, compact_stat->prev_compact_time) <
+        WT_BACKGROUND_COMPACT_SLEEP_TIME) {
+        compact_stat->skip_count++;
+        conn->background_compact.files_skipped++;
+        return (false);
+    }
+
+    /* If the last compaction pass was less successful than the average. Skip it for some time. */
+    if (compact_stat->bytes_rewritten < conn->background_compact.bytes_rewritten_ema &&
+      WT_CLOCKDIFF_SEC(cur_time, compact_stat->prev_compact_time) <
+        WT_BACKGROUND_COMPACT_SLEEP_TIME) {
+        compact_stat->skip_count++;
+        conn->background_compact.files_skipped++;
+        return (false);
+    }
+
+    return (true);
+}
+
+/*
+ * __background_compact_start --
+ *     Pre-fill compact related statistics for the given file.
+ */
+static int
+__background_compact_start(
+  WT_SESSION_IMPL *session, const char *uri, int64_t id, WT_BACKGROUND_COMPACT_STAT **compact_stat)
+{
+    WT_BACKGROUND_COMPACT_STAT *temp_compact_stat;
+    WT_BM *bm;
+    WT_DECL_RET;
+
+    bm = S2BT(session)->bm;
+
+    temp_compact_stat = __background_compact_get_stat(session, uri, id);
+
+    /* If the table is not in the list, allocate a new entry and insert it. */
+    if (temp_compact_stat == NULL) {
+        WT_ERR(__wt_calloc_one(session, &temp_compact_stat));
+        WT_ERR(__wt_strdup(session, uri, &temp_compact_stat->uri));
+        __background_compact_list_insert(session, temp_compact_stat);
+    }
+
+    /* Fill starting information prior to running compaction. */
+    WT_ERR(bm->size(bm, session, &temp_compact_stat->start_size));
+    temp_compact_stat->prev_compact_time = __wt_clock(session);
+
+    *compact_stat = temp_compact_stat;
+
+    return (0);
+
+err:
+    __wt_free(session, temp_compact_stat);
```

Comment:
It may not be possible at the moment, but I suspect this should remove the `compact_stat` from the hash before freeing the data structure.

Reply:
That should be handled by `__background_compact_list_cleanup` in the error path of `__compact_server`